pkg/scheduler/plugins/reservation/transformer.go

```diff
@@ -22,210 +22,361 @@ import (
-		count := 0
-		rCache := getReservationCache()
-		hasAllocatedResource := false
-		for _, obj := range rOnNode {
-			r, ok := obj.(*schedulingv1alpha1.Reservation)
-			if !ok {
-				klog.V(5).Infof("unable to convert to *schedulingv1alpha1.Reservation, obj %T", obj)
+		rOnNode := pl.reservationCache.listReservationInfosOnNode(node.Name)
+		for _, rInfo := range rOnNode {
+			if !reservationutil.IsReservationAvailable(rInfo.reservation) {
 				continue
 			}
-			rInfo := rCache.GetInCache(r)
-			if rInfo == nil {
-				rInfo = newReservationInfo(r)
-			}
-			// only count available reservations, ignore succeeded ones
-			if !reservationutil.IsReservationAvailable(rInfo.Reservation) {
+
+			// In this case, the Controller has not yet updated the status of the Reservation to Succeeded,
+			// but in fact it can no longer be used for allocation. So it's better to skip first.
+			if rInfo.reservation.Spec.AllocateOnce && len(rInfo.pods) > 0 {
 				continue
 			}
 
-			if matchReservation(pod, rInfo) {
-				matchedCache.Add(r)
-				count++
+			if !isReservedPod && reservationutil.MatchReservationOwners(pod, rInfo.reservation) {
+				matched = append(matched, rInfo)
 			} else {
-				if len(rInfo.Reservation.Status.CurrentOwners) > 0 {
-					hasAllocatedResource = true
-					resourceNeedUnreserve = quotav1.Add(resourceNeedUnreserve, rInfo.Reservation.Status.Allocated)
+				if len(rInfo.allocated) > 0 {
+					unmatched = append(unmatched, rInfo)
+				}
+				if !isReservedPod {
+					klog.V(6).InfoS("got reservation on node does not match the pod", "reservation", klog.KObj(rInfo.reservation), "pod", klog.KObj(pod))
 				}
-				klog.V(6).InfoS("got reservation on node does not match the pod",
-					"reservation", klog.KObj(r), "pod", klog.KObj(pod), "reason",
-					dumpMatchReservationReason(pod, newReservationInfo(r)))
 			}
 		}
-		if hasAllocatedResource {
+
+		// Most scenarios do not have reservations. It is better to check if there is a reservation
+		// before deciding whether to add a lock update, which can reduce race conditions.
+		if len(matched) > 0 || len(unmatched) > 0 {
 			lock.Lock()
-			allocatedResource[node.Name] = resourceNeedUnreserve
+			if len(matched) > 0 {
+				state.matched[node.Name] = matched
+			}
+			if len(unmatched) > 0 {
+				state.unmatched[node.Name] = unmatched
+			}
 			lock.Unlock()
 		}
-		if count <= 0 { // no reservation matched on this node
+
+		if len(matched) == 0 {
 			return
 		}
 
-		// NOTE: when the pod can allocate any reservation on the node, we should alter the nodeInfo snapshot to skip
-		//  the affinity/anti-affinity/topo constrains filtering in InterPodAffinity and PodTopologySpread plugins.
-		preparePreFilterNodeInfo(handle, nodeInfo, pod, matchedCache)
-		klog.V(4).InfoS("BeforePreFilter get matched reservations", "pod", klog.KObj(pod),
-			"node", node.Name, "count", count)
+		restorePVCRefCounts(pl.handle.SharedInformerFactory(), nodeInfo, pod, matched)
```

Comment:
why not restore PVCRefcounts afterPrefilter?

Reply:
There are related comments in this function, mainly to solve the problem of VolumeRestriction.